### PR TITLE
[HDRP] Fullscreen subtarget feature

### PR DESCRIPTION
**Pull request by alelievr**
### Purpose of this PR
⚠️ This PR depends on https://github.com/Unity-Technologies/Graphics/pull/6110 (so it needs to be merged after)

Included in this PR:
- Fullscreen sub-target for HDRP

image

- HD Scene depth node to sample the depth with mips
HDSceneDepth

- HD Sample Buffers node, allows to access normal, roughness, motion vectors, and post-process input in ShaderGraph.
image

- Full support of stencil in custom passes (in DrawRenderers and Fullscreen custom passes)

image
image

If you have any questions regarding this feature, you can ask in [#devs-fullscreen-master-node](https://unity.slack.com/archives/C029VTRHMV0).

---
### Testing status
Fullscreen target test doc: https://confluence.unity3d.com/display/HDRP/Fullscreen+Target

Automated tests have also been added for basic features of the fullscreen SG with custom post-process and custom passes (tests cover alpha blending, stencil, and fullscreen UVs).

### Note for QA:
This work is split into 3 PRs for ease of reviewing. If you want to test everything (URP, HDRP, and builtin implementation) I'd recommend to use the [sg/fullscreen-target](https://github.com/Unity-Technologies/Graphics/tree/sg/fullscreen-target) branch

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/Material/Fullscreen/HDFullscreenFunctions.hlsl [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
+// For now we fake the GI sampling in fullscreen.
+float3 SampleBakedGI(float3 positionRWS, float3 normalWS, float2 uvStaticLightmap, float2 uvDynamicLightmap)
```
Comment (kecho) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Fullscreen/HDFullscreenFunctions.hlsl:3]: maybe better to not expose functions that dont have functionality?

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+                    s.AppendLine("#include \"Packages/com.unity.render-pipelines.high-definition/Runtime/Material/NormalBuffer.hlsl\"");
+                    s.AppendLine("#include \"Packages/com.unity.render-pipelines.high-definition/Runtime/Material/Builtin/BuiltinData.hlsl\"");
+
+                    s.AppendLine("$precision{1} {0}($precision2 uv, SamplerState samplerState)", GetFunctionName(), channelCount);
+                    using (s.BlockScope())
+                    {
+                        switch (bufferType)
+                        {
+                            case BufferType.NormalWorldSpace:
+                                s.AppendLine("uint2 pixelCoords = uint2(uv * _ScreenSize.xy);");
```
Comment (kecho) [inline com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs:148]: Consider wrapping this into simpler functions

**Review by kecho — COMMENTED** [review]
Have you tested this with DRS & DLSS? 
Left some small comments, nothing blocker. Nice job!

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs [hunk 3]
```diff
@@ -40,6 +40,9 @@ public partial class HDRenderPipeline
                 // Be careful, ComputePackedMipChainInfo needs the render texture size and not the viewport size. Otherwise it would compute the wrong size.
                 hdCamera.depthBufferMipChainInfo.ComputePackedMipChainInfo(RTHandles.rtHandleProperties.currentRenderTargetSize);
 
+                // Bind the depth pyramid offset info for the HDSceneDepth node in ShaderGraph. This can be used by users in custom passes.
+                Shader.SetGlobalBuffer(HDShaderIDs._DepthPyramidMipLevelOffsets, hdCamera.depthBufferMipChainInfo.GetOffsetBufferData(m_DepthPyramidMipLevelOffsetsBuffer));
```
[the remark this answers is not in the document]
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs:44]: Yeah, in this part of the core we're in the camera-specific section (checked with Julien)

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs (hunk 2 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs:148]: In C# or the generated code?

File: com.unity.render-pipelines.high-definition/Runtime/Material/Fullscreen/HDFullscreenFunctions.hlsl (hunk 1 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/Material/Fullscreen/HDFullscreenFunctions.hlsl:3]: Yeah, I've marked the "BakedGINode" in ShaderGraph incompatible with the fullscreen target

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs (hunk 2 above)
Comment (kecho) [reply com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs:121]: I was hoping we could have something like a straight hlsl files we could include here, keep this complex code in such files, and just call the functions from here. If its too hard to do dont worry about it.

**Review by kecho — APPROVED** [review]

File: TestProjects/HDRP_Tests/ProjectSettings/EditorBuildSettings.asset [hunk 4]
```diff
@@ -815,4 +818,4 @@ EditorBuildSettings:
   - enabled: 1
     path: Assets/GraphicTests/Scenes/9x_Other/9910_GlobalMipBias.unity
     guid: 3f9d18568d46df044847954b624b6152
-
+  m_configObjects: {}
```
Comment (sebastienlagarde) [inline TestProjects/HDRP_Tests/ProjectSettings/EditorBuildSettings.asset:821]: side note: last time I see someone touching this it cause some failure on Yamato, so let's monitor yamato to be sure it is green

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+                                s.AppendLine("NormalData normalData;");
+                                s.AppendLine("DecodeFromNormalBuffer(pixelCoords, normalData);");
+                                s.AppendLine("float depth = LoadCameraDepth(pixelCoords);");
+                                s.AppendLine("return depth > 0 ? normalData.normalWS : 0;");
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs:125]: Remark, this test assume invert Z (which is the case for hdrp) but to have in mind

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs:125]: we should rather abstract with an "IsSky()" instead

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs:125]: Also generally speaking, I think it would be important in shader graph to have a way to know if a pixel from a target( normal / roughness) is "valid". As here users have no way to know it other than redoing this exact same test with depth

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs:125]: hmm... maybe currently this node (HDSampleBuffer), should have an extra output to say that input is valid or not

File: com.unity.render-pipelines.high-definition/Editor/RenderPipeline/CustomPass/FullScreenCustomPassDrawer.cs [hunk 6]
```diff
@@ -67,21 +76,147 @@ protected override void DoPassGUI(SerializedProperty customPass, Rect rect)
...
                 }
             }
         }
 
+        bool DoesWriteMaskContainsReservedBits(int writeMask)
+        {
+            if (targetDepthBuffer == CustomPass.TargetBuffer.Custom)
+                return false;
+
+            return ((writeMask & (int)~(UserStencilUsage.UserBit0 | UserStencilUsage.UserBit1)) != 0);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/RenderPipeline/CustomPass/FullScreenCustomPassDrawer.cs:136]: we should create a mask for all reserverd bit. Otherwise if in the future we change it we will need to edit all place like that. i.e do return ((writeMask & (int)~(UserStencilUsage.UserBitMask)

File: com.unity.render-pipelines.high-definition/Runtime/Debug/DebugLightVolumes.shader [hunk 7]
```diff
@@ -68,6 +69,7 @@ Shader "Hidden/HDRP/DebugLightVolumes"
             HLSLPROGRAM
             #pragma vertex vert
             #pragma fragment frag
+            #pragma only_renderers d3d11 playstation xboxone xboxseries vulkan metal switch
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Debug/DebugLightVolumes.shader:72]: Can you do a separate PR for this one as it will need to be backport, thanks

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugLightVolumes.shader:72]: (and same for other like this)

File: com.unity.render-pipelines.high-definition/Runtime/Material/Fullscreen/HDFullscreenFunctions.hlsl [hunk 8]
```diff
@@ -0,0 +1,6 @@
...
+{
+    return 0;
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Fullscreen/HDFullscreenFunctions.hlsl:5]: I am not sure to understand why we need a new file just for this? why is it here?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs:44]: But why this is not using a command buffer command? With this you can easily be out of sync

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs [hunk 9]
```diff
@@ -40,6 +40,9 @@ public partial class HDRenderPipeline
                 // Be careful, ComputePackedMipChainInfo needs the render texture size and not the viewport size. Otherwise it would compute the wrong size.
                 hdCamera.depthBufferMipChainInfo.ComputePackedMipChainInfo(RTHandles.rtHandleProperties.currentRenderTargetSize);
 
+                // Bind the depth pyramid offset info for the HDSceneDepth node in ShaderGraph. This can be used by users in custom passes.
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs:43]: But why this is not using a command buffer command? With this you can easily be out of sync

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs [hunk 10]
```diff
@@ -228,7 +228,7 @@ internal void ExecuteInternal(RenderGraph renderGraph, HDCamera hdCamera, Cullin
                         if (customPass.currentRenderTarget.colorBufferRG.IsValid() && customPass.injectionPoint == CustomPassInjectionPoint.AfterPostProcess)
                             ctx.cmd.SetGlobalTexture(HDShaderIDs._AfterPostProcessColorBuffer, customPass.currentRenderTarget.colorBufferRG);
 
-                        if (customPass.currentRenderTarget.motionVectorBufferRG.IsValid() && (customPass.injectionPoint == CustomPassInjectionPoint.BeforePostProcess || customPass.injectionPoint == CustomPassInjectionPoint.AfterPostProcess))
+                        if (customPass.currentRenderTarget.motionVectorBufferRG.IsValid() && (customPass.injectionPoint != CustomPassInjectionPoint.BeforeRendering))
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs:231]: is it a bugfix or just synatix sugar?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/DrawRenderersCustomPass.cs [hunk 11]
```diff
@@ -66,6 +66,46 @@ public enum ShaderPass
...
+
+        /// <summary>
+        /// Write mask of the stencil.
+        /// </summary>
+        public int stencilWriteMask = (int)(UserStencilUsage.UserBit0 | UserStencilUsage.UserBit1);
+
+        /// <summary>
+        /// Read mask of the stencil
+        /// </summary>
+        public int stencilReadMask = (int)(UserStencilUsage.UserBit0 | UserStencilUsage.UserBit1);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/DrawRenderersCustomPass.cs:87]: same remark than before about creating a reservert userBitMask. and IIRC there is one currently

**Review by sebastienlagarde — APPROVED** [review]

File: TestProjects/HDRP_Tests/ProjectSettings/EditorBuildSettings.asset (hunk 4 above)
Comment (alelievr) [reply TestProjects/HDRP_Tests/ProjectSettings/EditorBuildSettings.asset:821]: I can remove this change since we probably don't need it

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs (hunk 10 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs:231]: it's a bugfix

File: com.unity.render-pipelines.high-definition/Runtime/Debug/DebugLightVolumes.shader (hunk 7 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/Debug/DebugLightVolumes.shader:72]: https://github.com/Unity-Technologies/Graphics/pull/6204

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs (hunk 9 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.RenderGraph.cs:43]: No idea :) will fix

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs (hunk 10 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/RenderPass/CustomPass/CustomPass.cs:231]: https://github.com/Unity-Technologies/Graphics/pull/6205

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs (hunk 5 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/Nodes/HDSampleBufferNode.cs]: I've added an IsSky function and an extra output for the node that return the IsSky() value

File: com.unity.render-pipelines.high-definition/Editor/RenderPipeline/CustomPass/FullScreenCustomPassDrawer.cs (hunk 6 above)
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/RenderPipeline/CustomPass/FullScreenCustomPassDrawer.cs]: done

**Review by remi-chapelain — COMMENTED** [review]
Took over for while @TomasKiniulis was OoO. 

Here's what I've tested: 
- Created a brand new custom post process (simple thing that distorts UV when sampling PP inputs) with full screen shader master node to made sure I understood the feature.
- Did not managed to repro the issue mentionned earlier with DX12 (has been fixed) ✔️ 
- Metal issue is supposed to be fixed but can't verify (no hardware but trusting antoine on this) ✔️ 
- Tested all DRS algorithms in software / hardware (dx12) ⚠️ (some issues)
- Pair-tested in Built-in with @alelievr using two different methods. Working as expected ✔️ 
- --- First one is by using post process package in built in and feeding it a material made with shader graph full screen master node. ✔️ 
- --- Second one is by overriding the render of the camera using the "OnRenderImage()" function by blit-ing the source texture into the destination using a temporary buffer in between to avoid read/write issues. ✔️ 

For the details about issues for DRS, what I've found is depending on what is sampled in the HDSampleBuffer node, FSR gives slightly darker results but ONLY when roughness or WSnormals are sampled. 
Does NOT seem to repro when using IsSky, PostProcess or MotionVectors for some reasons. 
https://user-images.githubusercontent.com/57442369/140384575-182694b6-44c8-4076-9b1e-9d11ff9677b1.mp4
